### PR TITLE
Added upper-bound size limit to the kafka producer

### PR DESCRIPTION
**Pull request by rogaha**

### CODE REVIEW

File: kafka/producer/base.py [hunk 1]
```diff
@@ -127,9 +140,7 @@ def __init__(self, client, async=False,
         self.ack_timeout = ack_timeout
         self.stopped = False
 
-        if codec is None:
-            codec = CODEC_NONE
-        elif codec not in ALL_CODECS:
+        if codec not in ALL_CODECS:
```
Comment (bcwalrus) [inline kafka/producer/base.py]: Would revert to original. And leave L116 as original as well.

File: kafka/producer/base.py [hunk 2]
```diff
@@ -60,11 +62,13 @@ def _send_upstream(queue, client, codec, batch_time, batch_size,
             # Adjust the timeout to match the remaining period
             count -= 1
             timeout = send_at - time.time()
-            msgset[topic_partition].append((msg, key))
+            msgset[topic_partition].append(msg)
```
Comment (bcwalrus) [inline kafka/producer/base.py]: Why do we want to change this? Why not keep the dictionary key as a tuple of `(msg, key)`?

File: kafka/producer/base.py [hunk 3]
```diff
@@ -35,7 +38,6 @@ def _send_upstream(queue, client, codec, batch_time, batch_size,
     a specified timeout and send them upstream to the brokers in one
     request
     """
-    stop = False
```
Comment (bcwalrus) [inline kafka/producer/base.py:38]: Looks like you removed an unused variable. Don't.

The business of forking is to minimize the delta, so that we only add what's absolutely necessary for the feature we want.

File: kafka/producer/base.py [hunk 4]
```diff
@@ -27,6 +27,9 @@
 
 STOP_ASYNC_PRODUCER = -1
 
+# unlimited
+ASYNC_QUEUE_MAXSIZE = 0
```
Comment (bcwalrus) [inline kafka/producer/base.py]: Would default to some large limit, e.g. 65536.

File: kafka/producer/base.py [hunk 5]
```diff
@@ -168,12 +179,10 @@ def send_messages(self, topic, partition, *msg):
         @param: *msg, one or more message payloads -- type bytes
         @returns: ResponseRequest returned by server
         raises on error
-
```
Comment (bcwalrus) [inline kafka/producer/base.py:171]: Nit: Let's reduce the white space changes as well. It makes future merging harder.

File: kafka/producer/base.py [hunk 6]
```diff
@@ -190,26 +199,27 @@ def _send_messages(self, topic, partition, *msg, **kwargs):
         if any(not isinstance(m, six.binary_type) for m in msg):
             raise TypeError("all produce message payloads must be type bytes")
 
-        # Raise TypeError if topic is not encoded as bytes
-        if not isinstance(topic, six.binary_type):
-            raise TypeError("the topic must be type bytes")
-
```
Comment (bcwalrus) [inline kafka/producer/base.py:196]: Why remove the check?

File: kafka/producer/base.py [hunk 7]
```diff
@@ -190,26 +199,27 @@ def _send_messages(self, topic, partition, *msg, **kwargs):
...
+                try:
+                    item = (TopicAndPartition(topic, partition), m, key)
+                    self.queue.put_nowait(item)
+                except Full:
+                    log.exception('Async queue is full')
+                    raise
             resp = []
         else:
-            messages = create_message_set([(m, key) for m in msg], self.codec, key)
+            messages = create_message_set(msg, self.codec, key)
```
Comment (bcwalrus) [inline kafka/producer/base.py]: Would keep the original.

File: kafka/producer/simple.py [hunk 8]
```diff
@@ -41,19 +42,24 @@ class SimpleProducer(Producer):
...
                  req_acks=Producer.ACK_AFTER_LOCAL_WRITE,
                  ack_timeout=Producer.DEFAULT_ACK_TIMEOUT,
-                 codec=None,
+                 codec=CODEC_NONE,
                  batch_send=False,
                  batch_send_every_n=BATCH_SEND_MSG_COUNT,
                  batch_send_every_t=BATCH_SEND_DEFAULT_INTERVAL,
-                 random_start=True):
+                 random_start=True,
+                 maxsize=0):
```
Comment (bcwalrus) [inline kafka/producer/simple.py]: You probably mean `maxsize=None` here, and let the `Producer` class determine the default.

Right now you have the default in 2 places, both here and in `ASYNC_QUEUE_MAXSIZE`.

File: kafka/producer/simple.py [hunk 9]
```diff
@@ -63,7 +69,7 @@ def _next_partition(self, topic):
             # Randomize the initial partition that is returned
             if self.random_start:
                 num_partitions = len(self.client.get_partition_ids_for_topic(topic))
-                for _ in xrange(random.randint(0, num_partitions-1)):
+                for _ in xrange(random.randint(0, num_partitions - 1)):
```
Comment (bcwalrus) [inline kafka/producer/simple.py]: Would undo the whitespace change. Goal is to reduce delta when forking.

File: kafka/producer/simple.py [hunk 10]
```diff
@@ -78,4 +84,4 @@ def send_messages(self, topic, *msg):
         )
 
     def __repr__(self):
-        return '<SimpleProducer batch=%s>' % self.async
+        return u'<{0} batch={1}>'.format(self.__class__.__name__, self.async)
```
Comment (bcwalrus) [inline kafka/producer/simple.py]: Would use the original, since that works.

File: kafka/producer/base.py [hunk 11]
```diff
@@ -108,15 +113,23 @@ class Producer(object):
     def __init__(self, client, async=False,
                  req_acks=ACK_AFTER_LOCAL_WRITE,
                  ack_timeout=DEFAULT_ACK_TIMEOUT,
-                 codec=None,
+                 codec=CODEC_NONE,
                  batch_send=False,
                  batch_send_every_n=BATCH_SEND_MSG_COUNT,
-                 batch_send_every_t=BATCH_SEND_DEFAULT_INTERVAL):
+                 batch_send_every_t=BATCH_SEND_DEFAULT_INTERVAL,
+                 maxsize=ASYNC_QUEUE_MAXSIZE):
```
Comment (bcwalrus) [inline kafka/producer/base.py]: Would put `maxsize=None` here. Then in the body below, you can do:

```
if maxsize is None:
    maxsize = ASYNC_QUEUE_MAXSIZE
```

That allows `SimpleProducer` to pass in `None` and get the default behaviour.

File: test/test_producer.py [hunk 12]
```diff
@@ -40,3 +41,39 @@ def partitions(topic):
...
+        queue_size = 2
+        producer = Producer(MagicMock(), batch_send=True, maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        with self.assertRaises(Full):
+            message_list = [message] * (queue_size + 1)
+            producer.send_messages(topic, partition, *message_list)
```
Comment (bcwalrus) [inline test/test_producer.py:59]: Would check that the first 2 messages got into the queue.

File: test/test_producer.py [hunk 13]
```diff
@@ -40,3 +41,39 @@ def partitions(topic):
...
+        queue_size = 4
+        producer = Producer(MagicMock(), async=True, maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        message_list = [message] * (queue_size / 2 + 1)
+        resp = producer.send_messages(topic, partition, *message_list)
+        self.assertEqual(type(resp), list)
```
Comment (bcwalrus) [inline test/test_producer.py:92]: I think `len(resp) == len(message_list)`. Could you assert on that?

What's the state of the queue at the end here? Could you assert that as well?

File: kafka/producer/base.py (hunk 4 above)
Comment (rogaha) [inline kafka/producer/base.py]: @bcwalrus a large limit would be same as unlimited, isn't it? The python queue handles `zero` as unlimited.

File: kafka/producer/base.py (hunk 3 above)
Comment (rogaha) [inline kafka/producer/base.py:38]: Ok, got it!

File: kafka/producer/base.py (hunk 2 above)
Comment (rogaha) [inline kafka/producer/base.py]: I think it was there because of the version we were using `0.9.3`.  It changed with the current master.

File: kafka/producer/base.py (hunk 6 above)
Comment (rogaha) [inline kafka/producer/base.py:196]: It was re-introduced below, it must be some merging conflicts, I'll revert it.

File: kafka/producer/simple.py [hunk 14]
```diff
@@ -6,14 +6,14 @@
...
-from six.moves import xrange
+from kafka.protocol import CODEC_NONE
 
 from .base import (
     Producer, BATCH_SEND_DEFAULT_INTERVAL,
     BATCH_SEND_MSG_COUNT
 )
 
-log = logging.getLogger("kafka")
+log = logging.getLogger(__name__)
```
Comment (marcusmartins) [inline kafka/producer/simple.py]: As this is now a library again, we should revert the change.

File: kafka/producer/simple.py [hunk 15]
```diff
@@ -6,14 +6,14 @@
 
 from itertools import cycle
 
-from six.moves import xrange
```
Comment (marcusmartins) [inline kafka/producer/simple.py:9]: We should revert this change as well.

File: kafka/producer/base.py [hunk 16]
```diff
@@ -190,26 +199,27 @@ def _send_messages(self, topic, partition, *msg, **kwargs):
...
-            messages = create_message_set([(m, key) for m in msg], self.codec, key)
+            messages = create_message_set(msg, self.codec, key)
             req = ProduceRequest(topic, partition, messages)
             try:
                 resp = self.client.send_produce_request([req], acks=self.req_acks,
                                                         timeout=self.ack_timeout)
-            except Exception:
-                log.exception("Unable to send messages")
+            except:
+                log.exception('Unable to send messages. Topics: {0} Number of msgs: {1}'.format(topic, len(messages)))
```
Comment (marcusmartins) [inline kafka/producer/base.py]: I think we can keep the original message here.

File: kafka/producer/simple.py [hunk 17]
```diff
@@ -41,19 +42,24 @@ class SimpleProducer(Producer):
...
                                              ack_timeout, codec, batch_send,
                                              batch_send_every_n,
-                                             batch_send_every_t)
+                                             batch_send_every_t,
+                                             maxsize)
 
     def _next_partition(self, topic):
+        # for each message sent a partition is chosen, if random_start is True
+        # a random partition is selected, otherwise the next partition in the cycle
+        # is picked up
```
Comment (marcusmartins) [inline kafka/producer/simple.py]: Along with bc comment, I think we can leave this comment out, we can reduce the size of the diff.

File: test/test_producer.py (hunk 13 above)
Comment (rogaha) [inline test/test_producer.py:92]: Sure, I'll extend the tests. Thanks @bcwalrus

File: kafka/producer/simple.py (hunk 14 above)
Comment (rogaha) [inline kafka/producer/simple.py]: Ok!

File: kafka/producer/base.py (hunk 16 above)
Comment (rogaha) [inline kafka/producer/base.py]: Ok

File: kafka/producer/simple.py (hunk 17 above)
Comment (rogaha) [inline kafka/producer/simple.py]: Ok

File: kafka/producer/base.py (hunk 4 above)
Comment (toli) [inline kafka/producer/base.py]: do you want to have it as _unlimited_ in this case? i thought the whole idea was to have a limit by default.. so maybe you should set it to 65536 in that case...

File: kafka/producer/base.py [hunk 18]
```diff
@@ -208,7 +230,7 @@ def _send_messages(self, topic, partition, *msg, **kwargs):
             try:
                 resp = self.client.send_produce_request([req], acks=self.req_acks,
                                                         timeout=self.ack_timeout)
-            except Exception:
+            except:
```
Comment (toli) [inline kafka/producer/base.py]: don't change it to a "naked exception", keep as it was `except Exception`...
minimize changes

File: test/test_producer.py (hunk 12 above)
Comment (rogaha) [inline test/test_producer.py:59]: Test added!

Comment (toli) [inline test/test_producer.py:59]: that's a good idea.. message_list is currently empty right now, so we could put some garbage in there with "A", "B" and verify that we have the right content

File: test/test_producer.py [hunk 19]
```diff
@@ -40,3 +41,45 @@ def partitions(topic):
...
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        with self.assertRaises(Full):
+            message_list = [message] * (queue_size + 1)
+            producer.send_messages(topic, partition, *message_list)
+        self.assertEqual(producer.queue.qsize(), queue_size)
+
+    def test_producer_async_queue_overfilled(self):
```
Comment (toli) [inline test/test_producer.py]: you could theoretically decompose the body into a helper method, since the only difference between this test and the one above, `test_producer_async_queue_overfilled_batch_send` is how the `Producer` is initialized.

not a big deal, just something to remember next time

File: kafka/producer/base.py (hunk 4 above)
Comment (rogaha) [inline kafka/producer/base.py]: I don't want to change the default behavior of the library. By default is `unlimited`, but you have the choice of setting up a limit. Does it make sense?

File: kafka/producer/base.py (hunk 18 above)
Comment (rogaha) [inline kafka/producer/base.py]: ok, good catch! :)

File: kafka/producer/base.py (hunk 4 above)
Comment (bcwalrus) [inline kafka/producer/base.py]: Large limit != unlimited. My point is that:
1. Callers should pass in the limit. If caller wants unbounded queue, it has to be explicitly set to 0.
2. And the library should be defensive, and impose a limit (though it's a large one).

File: test/test_producer.py (hunk 19 above)
Comment (rogaha) [inline test/test_producer.py]: Yeap! Got it!

File: test/test_producer.py [hunk 20]
```diff
@@ -3,6 +3,7 @@
 import logging
 
 from mock import MagicMock
+from multiprocessing.queues import Full
```
Comment (bcwalrus) [inline test/test_producer.py]: In `base.py`, you have `from queue import Full`, or `from Queue import Full`. Would stick with that.

File: kafka/producer/base.py (hunk 4 above)
Comment (toli) [inline kafka/producer/base.py]: agree with @bcwalrus 

i tend to fall into "prevent ability to do stupid things" category, so i'd prefer to put a limit by default, so that we prevent stupid/accidental behaviour in the future. as @bcwalrus says, if they truly want _unlimited_, they should pass in 0 in that case

File: test/test_producer.py [hunk 21]
```diff
@@ -40,3 +41,45 @@ def partitions(topic):
...
+        partition = 0
+        message = b'test-message'
+
+        acceptable_size = (queue_size / 2 + 1)
+
+        message_list = [message] * acceptable_size
+        resp = producer.send_messages(topic, partition, *message_list)
+        self.assertEqual(type(resp), list)
+        self.assertEqual(producer.queue.qsize(), acceptable_size)
+        self.assertEqual(producer.queue.full(), False)
```
Comment (bcwalrus) [inline test/test_producer.py]: Nit: s/assertEqual/assertFalse/

**Comment by bcwalrus** [comment]
Minor comments. LGTM otherwise. Thanks, @rogaha!

File: kafka/producer/base.py (hunk 4 above)
Comment (rogaha) [inline kafka/producer/base.py]: I got your point. It's fine to me, I was just trying to reduce the expected behavior of the current `master` branch. I'll set the suggested limit.

**Comment by rogaha** [comment]
PR updated! Thanks @bcwalrus

File: test/test_producer.py [hunk 22]
```diff
@@ -2,7 +2,12 @@
 
 import logging
 
+try:
+    from queue import Full
+except ImportError:
+    from Queue import Full
 from mock import MagicMock
+from multiprocessing.queues import Full
```
Comment (bcwalrus) [inline test/test_producer.py]: Remove?

Comment (rogaha) [inline test/test_producer.py]: Yeap! :)

File: kafka/producer/base.py [hunk 23]
```diff
@@ -75,8 +79,8 @@ def _send_upstream(queue, client, codec, batch_time, batch_size,
             client.send_produce_request(reqs,
                                         acks=req_acks,
                                         timeout=ack_timeout)
-        except Exception:
-            log.exception("Unable to send message")
+        except:
+            log.exception('Failed to send messages to Kafka. Batch size: {0}. Topics: {1}'.format(len(reqs), topics))
```
Comment (marcusmartins) [inline kafka/producer/base.py]: I thought I had left a comment on this. I would revert this to reduce the diff although this message is better.

File: kafka/producer/base.py [hunk 24]
```diff
@@ -111,12 +116,20 @@ def __init__(self, client, async=False,
...
-            assert batch_send_every_t > 0
+            if batch_send_every_n <= 0:
+                log.exception('Batch send message count lower than zero.')
+                raise ValueError
+            if batch_send_every_t <= 0:
+                log.exception('Batch send message interval lower than zero.')
+                raise ValueError
+            if maxsize < 0:
+                log.exception('Queue size upper bound lower than zero.')
+                raise ValueError
```
Comment (marcusmartins) [inline kafka/producer/base.py]: Someone might have made a comment but I would revert this back and add a additional assert for maxsize < 0.

Also, we should try to avoid adding a new exception to the class as it changes the contract.

File: kafka/producer/base.py (hunk 23 above)
Comment (rogaha) [inline kafka/producer/base.py]: Hum...I was in another place. I'll update it.

Comment (marcusmartins) [inline kafka/producer/base.py]: Thanks

File: kafka/producer/base.py (hunk 24 above)
Comment (rogaha) [inline kafka/producer/base.py]: Since it was suggested by @bcwalrus, I will defer this change to him. It would be in line with the `less diffs` mindset.

**Comment by marcusmartins** [comment]
@rogaha @bcwalrus 

I just left 2 comments, otherwise it LGTM. But I would squash the commits before merging in.

File: test/test_producer.py [hunk 25]
```diff
@@ -40,3 +44,45 @@ def partitions(topic):
...
+        producer = Producer(MagicMock(), async=True, maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        with self.assertRaises(Full):
+            message_list = [message] * (queue_size + 1)
+            producer.send_messages(topic, partition, *message_list)
+        self.assertEqual(producer.queue.qsize(), queue_size)
```
Comment (marcusmartins) [inline test/test_producer.py:76]: hey @rogaha, @toli found an issue with intermittent test failures if the producer.queue.qsize() has being drained before the assert. He's working on a patch

Comment (rogaha) [inline test/test_producer.py:76]: Yes, I saw that issue as well. I couldn't reproduce it consistently though.

Comment (toli) [inline test/test_producer.py:76]: i think i have a fix in https://github.com/docker-hub/kafka-python/pull/2, we mocked out the "sending part" - your queue would occasionally drain fast enough such that you wouldn't trigger a Full exception

Comment (rogaha) [inline test/test_producer.py:76]: Awesome! Thanks @toli

**Comment by toli** [comment]
now that tests are passing - LGTM

File: kafka/producer/base.py (hunk 24 above)
Comment (bcwalrus) [inline kafka/producer/base.py]: Agree with @marcusmartins.

When I was reviewing the code a month back, I couldn't tell what was originally there and what's by Roberto.

**Comment by rogaha** [comment]
PR squashed and updated /cc @bcwalrus @marcusmartins

**Comment by rogaha** [comment]
Merging
